beacon_node/network/src/sync/manager.rs

```diff
@@ -530,87 +579,83 @@ impl<T: BeaconChainTypes> SyncManager<T> {
-                            );
-                            self.network
-                                .downvote_peer(parent_request.last_submitted_peer.clone());
-                            self.request_parent(parent_request);
-                            break;
-                        }
-                        Err(e) => {
-                            parent_request.failed_attempts += 1;
-                            warn!(
-                                self.log, "Parent processing error";
-                                "error" => format!("{:?}", e)
-                            );
-                            self.network
-                                .downvote_peer(parent_request.last_submitted_peer.clone());
-                            self.request_parent(parent_request);
-                            break;
-                        }
+            // The last block in the queue is the only one that has not attempted to be processed yet.
+            //
+            // The logic here attempts to process the last block. If it can be processed, the rest
+            // of the blocks must have known parents. If any of them cannot be processed, we
+            // consider the entire chain corrupt and drop it, notifying the user.
+            //
+            // If the last block in the queue cannot be processed, we also drop the entire queue.
+            // If the last block in the queue has an unknown parent, we continue the parent
+            // lookup-search.
+
+            if let Some(chain) = self.chain.upgrade() {
+                let newest_block = parent_request
+                    .downloaded_blocks
+                    .pop()
+                    .expect("There is always at least one block in the queue");
+                match chain.process_block(newest_block.clone()) {
+                    Ok(BlockProcessingOutcome::ParentUnknown { .. }) => {
+                        // need to keep looking for parents
+                        // add the block back to the queue and continue the search
+                        parent_request.downloaded_blocks.push(newest_block);
+                        self.request_parent(parent_request);
+                        return;
                     }
-                } else {
-                    break;
-                }
-            }
-
-            if successes > 0 {
-                if let Some(chain) = self.chain.upgrade() {
-                    match chain.fork_choice() {
-                        Ok(()) => trace!(
-                            self.log,
-                            "Fork choice success";
-                            "block_imports" => successes,
-                            "location" => "parent request"
-                        ),
-                        Err(e) => error!(
-                            self.log,
-                            "Fork choice failed";
+                    Ok(BlockProcessingOutcome::Processed { .. })
```

Comment:
Could we add `BlockProcessingOutcome::WouldRevertFinalizedSlot` and `BlockProcessingOutcome::GenesisBlock` also as successful conditions (I missed these in my first attempt)